lib/regular_expression/nfa.rb

```diff
@@ -118,6 +123,13 @@ def build(root)
             start_state.unshift(Transition::NegativeLookahead.new(finish_state, node.value))
           when AST::CharacterRange
             start_state.unshift(Transition::Range.new(finish_state, node.left, node.right))
+            next unless flags.ignore_case?
+
+            new_left = char_in_other_case(node.left) || node.left
```

Comment:
So I don't think this quite works unfortunately. Since this actually moves the ranges around the unicode table, you can get some unexpected results.

Let's say we have a range of `"@"` (64) to `"C"` (67). This would change the range from `"@"` (64) to `"c"` (99). Unfortunately that includes everything from `"["` (91) to ``"`"`` (96).

tbh I'm not even sure if we're guaranteed that a capital character range will have 1-1 correspondence with a lowercase character range in terms of order of unicode representations. For example, check out the result of `(200..300).map { |v| [v].pack("U") }` ->

```ruby
=> ["È", "É", "Ê", "Ë", "Ì", "Í", "Î", "Ï", "Ð", "Ñ", "Ò", "Ó", "Ô", "Õ", "Ö", "×", "Ø", "Ù", "Ú", "Û", "Ü", "Ý", "Þ", "ß", "à", "á", "â", "ã", "ä", "å", "æ", "ç", "è", "é", "ê", "ë", "ì", "í", "î", "ï", "ð", "ñ", "ò", "ó", "ô", "õ", "ö", "÷", "ø", "ù", "ú", "û", "ü", "ý", "þ", "ÿ", "Ā", "ā", "Ă", "ă", "Ą", "ą", "Ć", "ć", "Ĉ", "ĉ", "Ċ", "ċ", "Č", "č", "Ď", "ď", "Đ", "đ", "Ē", "ē", "Ĕ", "ĕ", "Ė", "ė", "Ę", "ę", "Ě", "ě", "Ĝ", "ĝ", "Ğ", "ğ", "Ġ", "ġ", "Ģ", "ģ", "Ĥ", "ĥ", "Ħ", "ħ", "Ĩ", "ĩ", "Ī", "ī", "Ĭ"]
```

I think this might have to be solved at the bytecode level. I still think you should pass the information into the transition here though because you can technically pass flags at the individual group level, which we'll need to handle at some point.

Reply:
Oh, you're right. Thanks for pointing it out. I'll change the PR to draft and revisit when I get a chance.